src/net_builtin.c

```diff
@@ -894,40 +902,53 @@ static void mg_tcpip_poll(struct mg_tcpip_if *ifp, uint64_t now) {
 
 #if MG_ENABLE_TCPIP_PRINT_DEBUG_STATS
   if (expired_1000ms) {
-    const char *names[] = {"down", "up", "req", "ready"};
+    const char *names[] = {"down", "up", "req", "ip", "ready"};
     MG_INFO(("Status: %s, IP: %M, rx:%u, tx:%u, dr:%u, er:%u",
              names[ifp->state], mg_print_ip4, &ifp->ip, ifp->nrecv, ifp->nsent,
              ifp->ndrop, ifp->nerr));
   }
 #endif
+  // Handle gw ARP request timeout, order is important
+  if (expired_1000ms && ifp->state == MG_TCPIP_STATE_IP) {
+    ifp->state = MG_TCPIP_STATE_READY; // keep best-effort MAC
+    onstatechange(ifp);
+  }
```

Comment:
If our ARP request for the gw MAC times out. maybe there is no gw, go READY anyway, we'll keep our best-effort MACs (broadcast or DHCP server)